### DISCUSSION TITLE
How do I incorporate passport js for authentication?

### DISCUSSION THREAD

**Discussion by egithinji** [RobinTail/express-zod-api#921] [Q&A] [answered]
Any direction on how to incorporate passport js in express-zod-api? For google oauth2, I'm trying to do something like this. This is the callback/redirect uri endpoint in the oauth2 flow:

```
export const googleRedirectEndpoint = customEndpointsFactory
  .use(
    passport.authenticate("google", {
      successReturnToOrRedirect: "/",
      failureRedirect: "/login",
    })
  )
  .build({
    method: "get",
    input: inputValidator,
    output: outputValidator,
    handler: async ({ input, options, logger }) => {
      logger.debug("Options:", options);
     ///
    },
  });
  ```
  
 But it's getting stuck in the passport middleware, and doesn't do anything, neither does code execution get to the handler (on front end, it hangs on the part where user authorizes google sign in).

**Comment by RobinTail** [comment]
Hello @egithinji ,

could you please clarify: are you trying to integrate this solution:

https://www.passportjs.org/packages/passport-google-oauth2/

?

**Reply by egithinji** [reply]
Hi, yes that one! (or any of the strategies for oauth2 authentication).

**Comment by RobinTail** [comment]
@egithinji ,

Passport is a great library with a perfect abstraction levels. It's a good choice.

This is the actual function that `.authenticate()` creates:

https://github.com/jaredhanson/passport/blob/master/lib/middleware/authenticate.js#L70

They call it "middleware", however, it acts more like an endpoint handler, because in general scenario it does not call `next()`.
Instead, it calls `response.end()`:

https://github.com/jaredhanson/passport/blob/master/lib/middleware/authenticate.js#L333

Since `next()` is not called, your handler is never called as well.

Long story short: Passport takes the complete control over authentication routes, it's a not something that expected to be ran in between. Therefore, instead of trying to fit it into `EndpointsFactory::use()`, I'd suggest you to consider an approach of connecting `express-zod-api` to your own express app. This approach is explained in the documentation here:

https://github.com/RobinTail/express-zod-api#connect-to-your-own-express-app

Thus, you:
- create express app yourself, 
- follow the Passport integration instructions as they are for your authentication routes,
- use `attachRouting()` method of `express-zod-api` for your own routes only.
- start your app yourself.

This should work. I hope it will help you.
Let me know if that didn't work for some reason.

**Reply by RobinTail** [reply]
The only thing besides that you may need is the middleware for checking that the user is authenticated. Similar to this example:

https://github.com/mstade/passport-google-oauth2/blob/master/example/app.js#L130

You can rewrite it for using with express-zod-api like this:

```ts
const authCheckMiddleware = createMiddleware({
  input: z.object({}), // means no inputs
  middleware: async ({ request }) => ({
    isAuthenticated: request.isAuthenticated() // it becomes an option for your handler
  }),
});
```

- or you can throw in the middleware to keep invalidation in one place,
- or you can do `response.end()` — this should also prevent handler execution.

@egithinji

**Comment by egithinji** [comment]
Thank you so much @RobinTail  for taking your time to dig into this. Once I get a chance to try your solution I will get back to you on how it went.

**Comment by andremw** [comment, marked as answer]
I believe it's worth updating this one to highlight that it's possible to solve this problem by passing the provider attribute to the `addExpressMiddleware`, as described [here](https://ez.robintail.cz/v21.5.0/using-native-express-middlewares).
I just bumped into this need and this was how I solved it. 👌
